File: src/mantle/manipulation.ts
import { associateWith, filterEntries, flatten, isEmpty, omitKeys, transformValues } from '../core/objTransforms';
import {
  Id, IdMap, isBranching, isTerminal, NodeCommon, StringSlice, UnpositionedBranchingNode, UnpositionedNode,
  UnpositionedStrandedNode, UnpositionedTree
} from '../core/types';

export type NodeTransformFunc = (oldNode: UnpositionedNode) => UnpositionedNode;

type InsertedNodeCommon = NodeCommon & { targetParentId?: Id; }

type InsertedBranchingNode = InsertedNodeCommon & { targetChildIds: Id[]; };

type InsertedTerminalNode = InsertedNodeCommon & { targetSlice: StringSlice; };

export type InsertedNode = InsertedBranchingNode | InsertedTerminalNode;

const isIn = (nodes: IdMap<UnpositionedNode>) => (nodeId: Id) => nodes.hasOwnProperty(nodeId);

const slicesOverlap = ([start1, end1]: StringSlice, [start2, end2]: StringSlice) =>
  !(end1 < start1 || end2 < start2 || end1 <= start2 || end2 <= start1)

const descendantIds = (nodes: IdMap<UnpositionedNode>) => (node: UnpositionedBranchingNode): Id[] => {
  const directChildren = node.children.map(childId => nodes[childId]);
  const indirectDescendantIds = flatten(directChildren.filter(isBranching).map(descendantIds(nodes)));
  return [...node.children, ...indirectDescendantIds];
}

const descendantsOf = (nodes: IdMap<UnpositionedNode>) => (node: UnpositionedBranchingNode): IdMap<UnpositionedNode> =>
  associateWith(descendantIds(nodes)(node), childId => nodes[childId]);

const toStrandedNode = (oldNodes: IdMap<UnpositionedNode>) => (node: UnpositionedNode): UnpositionedStrandedNode => ({
  label: node.label,
  offset: node.offset,
  formerDescendants: isBranching(node) ? descendantsOf(oldNodes)(node) : undefined,
  formerSlice: isTerminal(node) ? node.slice : undefined,
  formerlyTriangle: isTerminal(node) ? node.triangle : undefined,
});

const insertNode =
  (insertedNode: InsertedNode) =>
  (nodeId: Id) =>
  (nodes: IdMap<UnpositionedNode>): IdMap<UnpositionedNode> => {
    let nodeMapWithNewNode = 'targetChildIds' in insertedNode
      ? {
        ...transformValues(nodes, node => isBranching(node)
          ? { ...node, children: node.children.filter(childId => !insertedNode.targetChildIds.includes(childId)) }
          : node),
        [nodeId]: {
          label: insertedNode.label,
          offset: { dTreeX: 0, dTreeY: 0 },
          children: insertedNode.targetChildIds,
        },
      }
      : {
        ...nodes,
        [nodeId]: {
          label: insertedNode.label,
          offset: { dTreeX: 0, dTreeY: 0 },
          slice: insertedNode.targetSlice,
          triangle: false,
        },
      };
    return insertedNode.targetParentId
      ? transformNode(node =>
        isBranching(node) ? { ...node, children: [...node.children, nodeId] } : node)(
          insertedNode.targetParentId)(nodeMapWithNewNode)
      : nodeMapWithNewNode;
  };

const transformNode =
  (transformFunc: NodeTransformFunc) =>
  (nodeId: Id) =>
  (nodes: IdMap<UnpositionedNode>): IdMap<UnpositionedNode> =>
    isIn(nodes)(nodeId)
      ? { ...nodes, [nodeId]: transformFunc(nodes[nodeId]) }
      : nodes;

const deleteNodes =
  (nodeIds: Id[]) =>
  (nodes: IdMap<UnpositionedNode>): IdMap<UnpositionedNode> => {
    if (isEmpty(nodeIds) || isEmpty(nodes)) return nodes;
    const filteredNodes = omitKeys(nodes, nodeIds);
    return transformValues(filteredNodes, node => {
      if (!isBranching(node)) return node;
      const filteredChildren = node.children.filter(childId => !nodeIds.includes(childId));
      if (node.children.length === filteredChildren.length) return node;
      if (isEmpty(filteredChildren)) return toStrandedNode(nodes)(node);
      return { ...node, children: filteredChildren };
    });
  };

/**
 * Inserts the given node into the tree, assigning it the given ID.
 */
export const insertNodeIntoTree =
  (node: InsertedNode) =>
  (newNodeId: Id) =>
  (tree: UnpositionedTree): UnpositionedTree => ({
    ...tree,
    nodes: insertNode(node)(newNodeId)(tree.nodes),
  });

/**
 * Transforms the node with the given ID using the given transform function
 * at any point in the given tree.
 */
export const transformNodeInTree =
  (transformFunc: NodeTransformFunc) =>
  (nodeId: Id) =>
  (tree: UnpositionedTree): UnpositionedTree => ({
    ...tree,
    nodes: transformNode(transformFunc)(nodeId)(tree.nodes),
  });

/**
 * Transforms all nodes in the given tree using the given transform function.
 */
export const transformAllNodesInTree =
  (transformFunc: NodeTransformFunc) =>
  (tree: UnpositionedTree): UnpositionedTree => ({
    ...tree,
    nodes: transformValues(tree.nodes, transformFunc),
  });

/**
 * Deletes the node with the given ID from the given tree.
 */
export const deleteNodesInTree =
  (nodeIds: Id[]) =>
  (tree: UnpositionedTree): UnpositionedTree => ({
    ...tree,
    nodes: deleteNodes(nodeIds)(tree.nodes),
  });

export const getParentNodeIdsInTree =
  (nodeIds: Id[]) =>
  (tree: UnpositionedTree): Id[] =>
    Object.keys(filterEntries(tree.nodes, ([_, node]) =>
      isBranching(node) && nodeIds.some(selectedNodeId => node.children.includes(selectedNodeId))));

export const getNodeIdsAssignedToSlice =
  (slice: StringSlice) =>
  (tree: UnpositionedTree): Id[] =>
    Object.keys(filterEntries(tree.nodes, ([_, node]) =>
      isTerminal(node) && slicesOverlap(slice, node.slice)));

File: npbloom-core/src/commonMain/kotlin/space/yuvalinguist/npbloom/content/unpositioned/Manipulation.kt
package space.yuvalinguist.npbloom.content.unpositioned

import space.yuvalinguist.npbloom.content.*

typealias NodeTransformFunc = (oldNode: UnpositionedNode) -> UnpositionedNode
typealias TreeTransformFunc = (oldTree: UnpositionedTree) -> UnpositionedTree

sealed interface InsertedNode : NodeBase {
    override val label: NodeLabel
    val targetParentId: Id?
}

internal data class InsertedBranchingNode(
    override val id: Id,
    override val label: NodeLabel,
    override val targetParentId: Id?,
    val targetChildIds: Set<Id>,
    override val yAlignMode: YAlignMode = YAlignMode.Bottom,
) : InsertedNode

internal data class InsertedTerminalNode(
    override val id: Id,
    override val label: NodeLabel,
    override val targetParentId: Id?,
    val targetSlice: StringSlice,
    val triangle: Boolean,
    override val yAlignMode: YAlignMode = YAlignMode.Bottom,
) : InsertedNode

internal fun EntitySet<UnpositionedNode>.descendantIds(node: UnpositionedBranchingNode): Set<Id> {
    val directChildren = node.children.map { this[it] }
    val indirectDescendantIds = directChildren.filterIsInstance<UnpositionedBranchingNode>().flatMap(::descendantIds)
    return node.children + indirectDescendantIds
}

internal fun EntitySet<UnpositionedNode>.descendantsOf(node: UnpositionedBranchingNode): EntitySet<UnpositionedNode> =
    EntitySet(descendantIds(node).map { this[it]!! })

/**
 * Returns the IDs of all nodes that are descendants of folded branching nodes.
 * These nodes should be hidden when rendering the tree.
 */
internal fun EntitySet<UnpositionedNode>.descendantsOfFolded(): EntitySet<UnpositionedNode> =
    filter { it is UnpositionedBranchingNode && it.folded }
        .flatMap { descendantIds(it as UnpositionedBranchingNode) }
        .toSet()
        .let { EntitySet(it.map { this[it]!! }) }

internal fun EntitySet<UnpositionedNode>.toStrandedNode(node: UnpositionedNode): UnpositionedStrandedNode =
    when (node) {
        is UnpositionedBranchingNode ->
            UnpositionedFormerlyBranchingNode(node.id, node.label, node.offset, descendantsOf(node))

        is UnpositionedTerminalNode ->
            UnpositionedFormerlyTerminalNode(node.id, node.label, node.offset, node.slice, node.triangle)

        else -> UnpositionedPlainStrandedNode(node.id, node.label, node.offset)
    }

internal fun EntitySet<UnpositionedNode>.unassignAsChildren(
    nodeIds: Set<Id>,
    node: UnpositionedNode
): UnpositionedNode {
    if (node !is UnpositionedBranchingNode) return node

    val filteredChildren = node.children - nodeIds
    if (node.children == filteredChildren) return node

    if (filteredChildren.isEmpty()) return toStrandedNode(node)

    return node.copy(children = filteredChildren)
}

internal fun EntitySet<UnpositionedNode>.insertNode(insertedNode: InsertedNode): EntitySet<UnpositionedNode> {
    val nodeMapWithNewNode: EntitySet<UnpositionedNode> = when (insertedNode) {
        is InsertedBranchingNode -> mapToNewEntitySet {
            if (it is UnpositionedBranchingNode) it.copy(children = it.children - insertedNode.targetChildIds)
            else it
        } + UnpositionedBranchingNode(
            insertedNode.id,
            insertedNode.label,
            TreeCoordsOffset.ZERO,
            insertedNode.targetChildIds,
        )

        is InsertedTerminalNode -> this + UnpositionedTerminalNode(
            insertedNode.id,
            insertedNode.label,
            TreeCoordsOffset.ZERO,
            insertedNode.targetSlice,
            insertedNode.triangle,
        )
    }
    return if (insertedNode.targetParentId != null) {
        nodeMapWithNewNode.transformNodes(setOf(insertedNode.targetParentId!!)) {
            if (it is UnpositionedBranchingNode) it.copy(children = it.children + insertedNode.id)
            else it
        }
    } else nodeMapWithNewNode
}

internal fun EntitySet<UnpositionedNode>.transformNodes(
    nodeIds: Set<Id>,
    transformFunc: NodeTransformFunc,
): EntitySet<UnpositionedNode> =
    nodeIds.fold(this) { transformedNodes, nodeId ->
        if (nodeId in this) transformedNodes + transformFunc(this[nodeId]!!)
        else transformedNodes
    }

internal fun EntitySet<UnpositionedNode>.deleteNodes(nodeIds: Set<Id>): EntitySet<UnpositionedNode> {
    if (nodeIds.isEmpty() || isEmpty()) return this
    val filteredNodes = filter { it.id !in nodeIds }
    return filteredNodes.mapToNewEntitySet { unassignAsChildren(nodeIds, it) }
}

internal fun EntitySet<UnpositionedTree>.transformTrees(
    treeIds: Set<Id>,
    transformFunc: TreeTransformFunc,
): EntitySet<UnpositionedTree> =
    treeIds.fold(this) { transformedTrees, treeId ->
        if (treeId in this) transformedTrees + transformFunc(this[treeId]!!)
        else transformedTrees
    }

internal fun EntitySet<UnpositionedTree>.deleteTrees(treeIds: Set<Id>): EntitySet<UnpositionedTree> =
    if (treeIds.isEmpty() || isEmpty()) this else filter { it.id !in treeIds }
